### ISSUE TITLE
Waiting for host Port 

### ISSUE THREAD

**Issue by Reiser-meta** [testcontainers/testcontainers-node#40]
Can you help me to solve this problem?? im runing version 2.0 .0 in a windows machine, yeah sad =/ but true. Here is the code: 

```
const containers = {}
describe(`Integration Testing ${__filename}`, () => {
  before(async done => {
    containers.mysqlContainer = await new GenericContainer('mysql')
      .withEnv('MYSQL_ROOT_PASSWORD', 'test')
      .withExposedPorts(3306)
    await containers.mysqlContainer.start()
    process.env.DB_HOST_TEST = `mysql://test:test@localhost:${containers.mysqlContainer.getMappedPort(3306)}`
    done()
  })
...
```
The container start, i can see it running docker ps but when i run npm test, it fails.

the error code: 

```
 Integration Testing 
  testcontainers INFO: Using Docker defaults +0ms
  testcontainers INFO: Creating container for image: mysql:latest +66ms
Redis conectado
  testcontainers INFO: Starting container with ID: 8e5214f4cabf114262aeb7a23811f01d0ab6caf04bb6448ae52021b4e2a8105d +173ms
  testcontainers DEBUG: Starting container health checks +1s
  testcontainers DEBUG: Waiting for host port :54979 +4ms
  testcontainers DEBUG: Waiting for internal port :3306 +5ms
```
And stops here, nothing responds. Any idea?

Thanks in advance.

**Comment by cristianrgreco** [comment]
@Reiser-meta It seems that the container isn't able to bind to port 3306. Do you have issues if you start a mysql container via the Docker CLI? If that works, could you exec onto the container and share the exit codes and output of the following commands?

```bash
/bin/sh -c "cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :0cea"
/bin/sh -c "nc -vz -w 1 localhost 3306"
/bin/bash -c "</dev/tcp/localhost/3306"
```

**Comment by Reiser-meta** [comment]
@cristianrgreco Ty for reply, the container is up, and i can access it via CLI. Im figuring out how to run this commands on windows and then i'll paste here!

**Comment by Reiser-meta** [comment]
if i run netstat -a which is equivalent to /proc/net/tcp, i can see the ports listening there.

```
  TCP    0.0.0.0:58929          GMEPN000676:0          LISTENING
  TCP    0.0.0.0:58973          GMEPN000676:0          LISTENING
  TCP    0.0.0.0:59498          GMEPN000676:0          LISTENING
  TCP    0.0.0.0:59861          GMEPN000676:0          LISTENING
  TCP    0.0.0.0:59978          GMEPN000676:0          LISTENING
  TCP    0.0.0.0:60571          GMEPN000676:0          LISTENING
```
all of them are mysql containers that was started by the code above. I think might be this ip 0.0.0.0, anyway, i´ve found this article, maybe it helps, https://github.com/docker/for-win/issues/1804

**Comment by cristianrgreco** [comment]
I meant to run those commands in the container itself. So for example if you have a container running with ID 123, you can run `docker exec -it 123 sh`, which will give you a shell onto the container. The Postgres container I assume is running Linux, so you should be able to run those three commands.

**Comment by Reiser-meta** [comment]
Oh , sorry for that...

```
cat: tcp6: No such file or directory
root@03d71264b4a3:/# cat /proc/net/tcp6 | awk '{print $2}' | grep -i :0cea
00000000000000000000000000000000:0CEA
root@03d71264b4a3:/#  nc -vz -w 1 localhost 3306
localhost [127.0.0.1] 3306 (?) open
```

**Comment by cristianrgreco** [comment]
So we can see that these commands return a success exit code:

```
root@03d71264b4a3:/# cat /proc/net/tcp6 | awk '{print $2}' | grep -i :0cea
00000000000000000000000000000000:0CEA

root@03d71264b4a3:/#  nc -vz -w 1 localhost 3306
localhost [127.0.0.1] 3306 (?) open
```

Which is what Testcontainers waits for before it determines that the internal port is bound, so I'm not sure why your logs show that Testcontainers is hanging after that. 

I think the default timeout is 60s, at which point Testcontainers will throw an error: "Port :${port} not bound after ${timeout}ms". Do you see this error? Are you able to step through the Testcontainers code and see what's stuck?

**Comment by Reiser-meta** [comment]
@cristianrgreco i can´t see this error, i´ve raised the timeout of mocha to 2 minutes in each step and still not seeing this error. I will try to debug Testcontainers code.

**Comment by Reiser-meta** [comment]
@cristianrgreco if i comment this line: const commandResults = yield Promise.all(commands.map(command => this.dockerClient.exec(this.container, command)));
it starts to work. I can´t figure out why.

**Comment by cristianrgreco** [comment]
@Reiser-meta Commenting that line would skip the check altogether, meaning that the container may be accessible to the app before it's ready to start accepting traffic.

If that line is the one that's hanging, that means that Testcontainers is waiting indefinitely for a result, which isn't ideal, perhaps we're missing a timeout, though you would still have an issue.

 There is a list of commands which are run to do the check, could you try removing then one by one and see if it's one in particular which is hanging?

```typescript
const commands = [
  ["/bin/sh", "-c", `cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :${portHex}`],
  ["/bin/sh", "-c", `nc -vz -w 1 localhost ${port}`],
  ["/bin/bash", "-c", `</dev/tcp/localhost/${port}`]
];
```

**Comment by Reiser-meta** [comment]
@cristianrgreco , none of this commands are hanging, even if i send an empty array of command, the application doesnt respond. i though it was netcat, because when i ran this commands in the container, ii had to install netcat to execute. But even if i comment this line, it stop to respond.

**Comment by cristianrgreco** [comment]
@Reiser-meta It doesn't make sense that commenting out ` await Promise.all(commands.map(command => this.dockerClient.exec(this.container, command)));` works, but setting commands to an empty array doesn't. If it's an empty array there should be no interaction with Docker, it'd essentially be `await Promise.all([])`.

What version of node are you using?

**Comment by Reiser-meta** [comment]
@cristianrgreco yeah =/ 10.16.3

**Comment by cristianrgreco** [comment]
> if i comment this line: const commandResults = yield Promise.all(commands.map(command => this.dockerClient.exec(this.container, command)));
> it starts to work. I can´t figure out why.

I don't think this can be true, if when we set `commands = []` and add the line back, it hangs. Could you perhaps check again? In theory running `await Promise.all([])` is causing the process to hang.

**Comment by cristianrgreco** [comment]
@Reiser-meta A similar issue came up and it turned out that it was the Docker system that was nearly out of space: https://github.com/testcontainers/testcontainers-node/issues/41.

Could you try a `docker system prune` and try again?

**Comment by cristianrgreco** [comment]
If you update to `2.1.0` you'll see a WARN logged if disk space is < 2GB

**Comment by RedGlow** [comment]
Sorry if I piggyback on this issue, but I'm finding myself in the same situation trying to run postgresql and apparently I have enough disk space.

I'm on Windows 10, node 10.14.2, Docker desktop community 2.1.0.5 (40693).

I reduced the command list to just one command (nc), and logged what happens in the exec call.
My exec function now looks like this:

```javascript
    exec(container, command) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log("Running", command);
            const exec = yield container.exec({
                cmd: command,
                attachStdout: true,
                attachStderr: true
            });
            console.log("Starting");
            const stream = yield exec.start();
            console.log("Got stream?", !!stream);
            const output = Buffer.concat(yield stream_to_array_1.default(stream)).toString();
            console.log("Got output?", !!output);
            const { exitCode } = yield exec.inspect();
            console.log("Got exit code", exitCode);
            return { output, exitCode };
        });
    }
```

And the output is:
```
Running [ '/bin/sh', '-c', 'nc -vz -w 1 localhost 5432' ]
Starting
Got stream? true
```

Which means the stream-to-array call blocks.

The command seems completed on the host machine:

```bash
bash-5.0# ps auxfwww
PID   USER     TIME  COMMAND
    1 postgres  0:00 postgres
   51 postgres  0:00 postgres: checkpointer process
   52 postgres  0:00 postgres: writer process
   53 postgres  0:00 postgres: wal writer process
   54 postgres  0:00 postgres: autovacuum launcher process
   55 postgres  0:00 postgres: stats collector process
   56 root      0:00 /bin/bash
   62 root      0:00 ps auxfwww
bash-5.0#
```

The command on its own, after all, returns immediately with success:
```bash
bash-5.0# /bin/sh -c 'nc -vz -w 1 localhost 5432'
localhost (127.0.0.1:5432) open
bash-5.0#
```

so I don't really know what's stuck in here. I'll keep investigating if I find something more.

PS: the "empty array command" situation earlier was blocking because it causes the next line `return commandResults.some(result => result.exitCode === 0);` to return false (empty array .some => false), and thus it goes in a loop.

**Comment by RedGlow** [comment]
If it can be useful, I turned on debug for docker-modem, getting:
```
Running commands: [ [ '/bin/sh', '-c', 'nc -vz -w 1 localhost 5432' ] ]
Running [ '/bin/sh', '-c', 'nc -vz -w 1 localhost 5432' ]
  modem Sending: { path:
   '/containers/4a96bd6ea0fc579c91ab21c802dc6be8afcfbfdf919c02983877c3a1b8c38efd/exec',
  method: 'POST',
  headers: { 'Content-Type': 'application/json', 'Content-Length': 93 },
  key: undefined,
  cert: undefined,
  ca: undefined,
  socketPath: '//./pipe/docker_engine' } +49ms
  modem Received: {"Id":"c097d2f08fd7bf352233560088243460a91e02e52167c48f7ed0ea8c52b2f4a3"}
 +17ms
Starting
  modem Sending: { path:
   '/exec/c097d2f08fd7bf352233560088243460a91e02e52167c48f7ed0ea8c52b2f4a3/start',
  method: 'POST',
  headers: { 'Content-Type': 'application/json', 'Content-Length': 2 },
  key: undefined,
  cert: undefined,
  ca: undefined,
  socketPath: '//./pipe/docker_engine' } +8ms
Got stream? true
```

**Comment by RedGlow** [comment]
(could this be the underlying bug? https://github.com/apocas/docker-modem/issues/83 )

**Comment by simonstadlinger** [comment]
Any further insights on this issue, @RedGlow @cristianrgreco?
I have the same problems with following setup
- _Image_ : rabbitmq:3-management
- _Docker_ :  Engine: 19.03.08, API 1.40, containerd: 1.2.13, runc: 1.0.0
- _Node_: 12.14.1
- _Windows_: Windows 10 Enterprise 1809

In my test setup I run following command
` rabbitmqbroker = await new GenericContainer("rabbitmq", "3-management")
            .withExposedPorts(15672, 5672)
            .start();`

Either way of listing the ports, the script stops at waiting for the second port to be bound.

`testcontainers INFO: Using Docker defaults +0ms
  testcontainers INFO: Creating container for image: rabbitmq:3-management +25ms
  testcontainers INFO: Starting container with ID: cea318a081ff326975eff149927d1fa30a1b069234486d262626d1818035827d +89ms
  testcontainers DEBUG: Waiting for container to be ready +374ms
  testcontainers DEBUG: Waiting for host port 51832 +3ms
  testcontainers DEBUG: Waiting for internal port 15672 +25ms
  testcontainers DEBUG: Waiting for host port 51833 +13ms
`

I tried the same setup on MacOS Catalina, where the second ports takes ca. 18secs but at least gets bound. 
Would be really glad, if this gets resolved.

**Comment by polaroi8d** [comment]
I have got the same issue. I tried to make some integration tests with Testcontainers with GenericContainer and our private docker images. TestContainers output is the same as above: 

```  testcontainers INFO: Using Docker defaults +0ms
  testcontainers INFO: Creating container for image: private repository url +27ms
  testcontainers INFO: Starting container with ID: 1248ff70a7b7544b248c784acbb8d3f642e865f9b95e8e85de52725b13273f36 +82ms
  testcontainers DEBUG: Waiting for container to be ready +414ms
  testcontainers DEBUG: Waiting for host port 59534 +1ms
  testcontainers DEBUG: Waiting for internal port 3506 +3ms
  testcontainers DEBUG: Waiting for host port 59533 +4ms``` 

Only changes in here that the mocha testing library failed after a second. So I got this error message: `Error: (HTTP code 409) unexpected - Container e6b83a85e2e3b421678e56d810e2b084584b1569aa0f74addeefef951e171caa is not running ` 

I tried with some official images like MySQL but it seems not works at all. 

Env: 
- Image: private
- Docker: Docker version 19.03.5, build 633a0ea
- Node: v13.10.1
- OS: MacOS Catalina 10.15.4

**Comment by cristianrgreco** [comment]
@RedGlow That's very interesting, so actually the health check command is returning immediately, which is expected, but waiting for the stream is blocking. Perhaps if the command doesn't return any output with the command it just blocks? I'll investigate this further.

**Comment by cristianrgreco** [comment]
@RedGlow @simonstadlinger @Reiser-meta Could you please try 2.8.1?

**Comment by AG-Teammate** [comment]
This is a crazy one, quick fix:

1. Make sure Docker for Windows is listening on the 2375 port. In my case this was broken: https://github.com/docker/for-win/issues/3286
Run: 
```
docker run -d --name docker-remote-api -p 2375:2375 -v /var/run/docker.sock:/var/run/docker.sock jarkt/docker-remote-api
```
2. Start testcontainers-node in TCP mode:
```
set DOCKER_HOST=tcp://localhost:2375
```

Details:
I was able to trace it to https://github.com/apocas/docker-modem/blob/c0eee3e9bd717a127c75cd6f04129e87da41a8ec/lib/modem.js#L286 where http.ClientRequest never receives a response or error event. It just hangs.
This may be related to quirky named pipes implementation, even .NET client developers had to workaround by sending zero-byte writes: https://github.com/dotnet/Docker.DotNet/blob/master/src/Docker.DotNet/DockerPipeStream.cs#L52
I have not looked further since the TCP option seems to work, even though the listening is broken on Windows without a workaround.

**Comment by cristianrgreco** [comment]
Hi @AG-Teammate 

Thanks very much for investigating that and sharing your findings. I will look at adding it to the README and seeing if there's something testcontainers can do to make it work.

Out of curiosity, did upgrading testcontainers to 2.8.1 not have any effect on the issue?

**Comment by AG-Teammate** [comment]
Hi @cristianrgreco 

> Out of curiosity, did upgrading testcontainers to 2.8.1 not have any effect on the issue?

I haven't tried older versions, but I think not since it's deep in docker-modem and possibly in Docker for Windows named pipes implementation.

**Comment by cristianrgreco** [comment]
Hi, could you please try 2.18.3? I've updated dockerode and docker-modem, with a major version update to docker-modem

**Comment by tomqwpl** [comment]
I am also encountering this.
My test code is basically the test from the README:
`const { GenericContainer } = require("testcontainers");

(async () => {
  const container = await new GenericContainer("redis")
    .withExposedPorts(6379)
    .start();

  await container.stop();
})();`

The container has started. Command output from various things run within the container:
`# cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :0cea
cat: /proc/net/tcp{,6}: No such file or directory`

`# nc -vz -w 1 localhost 3306
/bin/sh: 3: nc: not found`

 `# cat /dev/tcp/localhost/6379
cat: /dev/tcp/localhost/6379: No such file or directory
# ls /dev/tcp/localhost
ls: cannot access '/dev/tcp/localhost': No such file or directory
# ls /dev
core  fd  full  mqueue  null  ptmx  pts  random  shm  stderr  stdin  stdout  tty  urandom  zero`

The actual container I'm wanting to run I can use fine with the java testcontainers, but it fails with this version.

**Comment by cristianrgreco** [comment]
Hi @tomqwpl, does https://github.com/testcontainers/testcontainers-node/issues/40#issuecomment-623798900 help?

**Comment by tomqwpl** [comment]
Java testcontainers have already potentially dealt with this issue, see:
See https://github.com/testcontainers/testcontainers-java/pull/2195

Using tcp{,6} doesn't always work. On the face of it, I can't see how my test code would work for anyone, since the failure is a function of the redis image it's running.

**Comment by cristianrgreco** [comment]
@tomqwpl It works for me (and many others I guess) 🙂 Here's the output when I run your test code:

```
➜  test DEBUG=testcontainers node repro.js
  testcontainers INFO Using default Docker host: localhost, socket path: /var/run/docker.sock +0ms
  testcontainers INFO Creating container for image: redis:latest +27ms
  testcontainers INFO Starting container with ID: 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2 +49ms
  testcontainers DEBUG Waiting for container to be ready: 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2 +248ms
  testcontainers DEBUG Waiting for host port 56653 +1ms
  testcontainers DEBUG Waiting for internal port 6379 +2ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: M1:C 03 Sep 2020 15:40:28.709 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo +2ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: n1:C 03 Sep 2020 15:40:28.709 # Redis version=6.0.7, bits=64, commit=00000000, modified=0, pid=1, just started +0ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: �1:C 03 Sep 2020 15:40:28.709 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf +0ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: C1:M 03 Sep 2020 15:40:28.710 * Running mode=standalone, port=6379. +0ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: �1:M 03 Sep 2020 15:40:28.710 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128. +1ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: 21:M 03 Sep 2020 15:40:28.710 # Server initialized +0ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: �1:M 03 Sep 2020 15:40:28.710 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled. +0ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: ;1:M 03 Sep 2020 15:40:28.711 * Ready to accept connections +0ms
  testcontainers INFO Container is ready +212ms
  testcontainers INFO Stopping container with ID: 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2 +0ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: F1:signal-handler (1599147628) Received SIGTERM scheduling shutdown... +14ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: :1:M 03 Sep 2020 15:40:29.013 # User requested shutdown... +74ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: M1:M 03 Sep 2020 15:40:29.013 * Saving the final RDB snapshot before exiting. +0ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: 01:M 03 Sep 2020 15:40:29.015 * DB saved on disk +2ms
  testcontainers TRACE 779f08bd7501d805a758c677ad9c987d434417fc390719cfa74230738260dba2: F1:M 03 Sep 2020 15:40:29.015 # Redis is now ready to exit, bye bye... +0ms
```

Thanks for sharing that issue, I've just tried it locally and all the tests pass - are you able to try changing the testcontainers-node code locally and see if it fixes your issue? If so it'd be great to merge it in and get this resolved

**Comment by tomqwpl** [comment]
From the looks of it, the line
`const commandResults = yield Promise.all(commands.map((command) => this.dockerClient.exec(this.container, command)));`
never returns. So sounds like the same situation as the comment above.

I'm not going to use TCP mode, since that's insecure. I can make it work with a log message wait strategy for what I want to do.

**Comment by cristianrgreco** [comment]
@tomqwpl Were you able to try the fix in testcontainers/testcontainers-java#2195?

**Comment by tomqwpl** [comment]
@cristianrgreco, yes, that's what I was doing. But changing the commands it's running won't make any difference since fundamentally the commands never return. So the problem isn't (right now) the commands that are run, but the fact that running the commands hangs
